File: MarkdownEditor/MarkdownEditorTests/EditorViewControllerTests.m
//
//  EditorViewControllerTests.m
//  MarkdownEditorTests
//

#import <XCTest/XCTest.h>
#import "../MarkdownEditor/Sources/EditorViewController.h"
#import "../MarkdownEditor/Sources/Converter/TextConverter.h"
#import "../MarkdownEditor/Sources/Editor/EditorDialogPresenter.h"

@interface EditorViewController (Testing)
- (void)textDidEndEditing:(NSNotification *)notification;
- (void)textDidChange:(NSNotification *)notification;
- (BOOL)textView:(NSTextView *)textView shouldChangeTextInRanges:(NSArray<NSValue *> *)affectedRanges replacementStrings:(NSArray<NSString *> *)replacementStrings;
- (IBAction)boldButtonClicked:(NSButton *)sender;
- (IBAction)strikeThroughButtonClicked:(NSButton *)sender;
- (IBAction)italicButtonClicked:(NSButton *)sender;
- (IBAction)quoteButtonClicked:(NSButton *)sender;
- (IBAction)codeButtonClicked:(NSButton *)sender;
- (IBAction)insertLinkButtonClicked:(NSButton *)sender;
- (IBAction)listBulletedButtonClicked:(NSButton *)sender;
- (IBAction)listNumberedButtonClicked:(NSButton *)sender;
- (IBAction)newDocument:(id)sender;
- (IBAction)openDocument:(id)sender;
- (IBAction)saveDocument:(id)sender;
- (IBAction)saveDocumentAs:(id)sender;
- (EditorDialogPresenter *)dialogPresenter;
- (NSString *)documentBody;
- (void)replaceCharactersInRange:(NSRange)range withString:(NSString *)string;
- (void)applyFormat:(TextConverterFormat)format;
- (void)newFile;
- (BOOL)openFile;
- (BOOL)saveFile;
- (NSString *)defaultDirectoryPath;
- (void)promptForOpenURLWithCompletionHandler:(void (^)(BOOL result))handler;
- (void)promptForSaveURLWithCompletionHandler:(void (^)(BOOL result))handler;
- (void)reloadFileFromDiskIfNeeded;
@end

@interface StubEditorDialogPresenter : EditorDialogPresenter
@property (nonatomic) NSUInteger alertCallCount;
@property (nonatomic) NSUInteger confirmCallCount;
@property (nonatomic) NSUInteger openPanelCallCount;
@property (nonatomic) NSUInteger savePanelCallCount;
@property (nonatomic) NSModalResponse confirmationResponse;
@property (nonatomic, strong) NSURL *openURL;
@property (nonatomic, strong) NSURL *saveURL;
@property (nonatomic, copy) NSString *lastAlertTitle;
@property (nonatomic, copy) NSString *lastAlertMessage;
@end

@implementation StubEditorDialogPresenter

- (void)showAlertWithTitle:(NSString *)title message:(NSString *)message forWindow:(NSWindow *)window {
    self.alertCallCount += 1;
    self.lastAlertTitle = title;
    self.lastAlertMessage = message;
}

- (void)confirmDiscardingChangesForWindow:(NSWindow *)window
                        completionHandler:(void (^)(NSModalResponse returnCode))handler {
    self.confirmCallCount += 1;
    handler(self.confirmationResponse);
}

- (void)showOpenFilePanelForWindow:(NSWindow *)window
                       initialPath:(NSString *)initialPath
                 completionHandler:(void (^)(NSURL * _Nullable selectedURL))handler {
    self.openPanelCallCount += 1;
    handler(self.openURL);
}

- (void)showSaveFilePanelForWindow:(NSWindow *)window
                       initialPath:(NSString *)initialPath
                 completionHandler:(void (^)(NSURL * _Nullable selectedURL))handler {
    self.savePanelCallCount += 1;
    handler(self.saveURL);
}

@end

@interface TestableEditorViewController : EditorViewController
@property (nonatomic) NSUInteger newFileCallCount;
@property (nonatomic) NSUInteger saveFileCallCount;
@property (nonatomic) NSUInteger openFileCallCount;
@property (nonatomic) BOOL saveFileResult;
@end

@implementation TestableEditorViewController

- (void)newFile {
    self.newFileCallCount += 1;
}

- (BOOL)saveFile {
    self.saveFileCallCount += 1;
    return self.saveFileResult;
}

- (BOOL)openFile {
    self.openFileCallCount += 1;
    return YES;
}

@end

@interface EditorViewControllerTests : XCTestCase
@property (nonatomic, strong) NSTextView *textView;
@property (nonatomic, strong) StubEditorDialogPresenter *dialogPresenter;
@end

@implementation EditorViewControllerTests

- (void)configureController:(EditorViewController *)controller text:(NSString *)text {
    self.textView = [[NSTextView alloc] initWithFrame:NSMakeRect(0, 0, 320, 200)];
    self.dialogPresenter = [[StubEditorDialogPresenter alloc] init];
    self.textView.string = text;
    [controller setValue:self.textView forKey:@"textView"];
    [controller setValue:self.dialogPresenter forKey:@"dialogPresenter"];
}

- (EditorViewController *)makeControllerWithText:(NSString *)text {
    EditorViewController *controller = [[EditorViewController alloc] init];
    [self configureController:controller text:text];
    return controller;
}

- (TestableEditorViewController *)makeTestableControllerWithText:(NSString *)text {
    TestableEditorViewController *controller = [[TestableEditorViewController alloc] init];
    [self configureController:controller text:text];
    return controller;
}

- (NSString *)temporaryPathWithName:(NSString *)name {
    return [NSTemporaryDirectory() stringByAppendingPathComponent:name];
}

- (void)testDocumentMetadataUsesOutlets {
    EditorViewController *controller = [self makeControllerWithText:@"body"];
    XCTAssertEqualObjects([controller documentBody], @"body");
}

- (void)testDocumentMetadataFallsBackWhenOutletsAreMissing {
    EditorViewController *controller = [[EditorViewController alloc] init];

    XCTAssertEqualObjects([controller documentBody], @"Unknown");
}

- (void)testSetRepresentedObjectIsAccepted {
    EditorViewController *controller = [self makeControllerWithText:@"body"];
    NSObject *object = [[NSObject alloc] init];

    controller.representedObject = object;

    XCTAssertEqual(controller.representedObject, object);
}

- (void)testNewFileResetsContentPathAndDirtyState {
    EditorViewController *controller = [self makeControllerWithText:@"old"];
    [controller setValue:@"/tmp/example.md" forKey:@"filePath"];
    [controller setValue:@YES forKey:@"dirty"];

    [controller newFile];

    XCTAssertEqualObjects(self.textView.string, @"New File");
    XCTAssertNil([controller valueForKey:@"filePath"]);
    XCTAssertFalse([[controller valueForKey:@"dirty"] boolValue]);
}

- (void)testOpenFileRejectsMissingPathsAndDirectories {
    EditorViewController *controller = [self makeControllerWithText:@"old"];

    XCTAssertFalse([controller openFile]);

    [controller setValue:@"/tmp/markdown-editor-missing.md" forKey:@"filePath"];
    XCTAssertFalse([controller openFile]);

    [controller setValue:NSTemporaryDirectory() forKey:@"filePath"];
    XCTAssertFalse([controller openFile]);
}

- (void)testOpenFileLoadsMarkdownAndClearsDirtyState {
    EditorViewController *controller = [self makeControllerWithText:@"old"];
    NSString *path = [self temporaryPathWithName:@"markdown-editor-open.md"];
    [@"# Opened" writeToFile:path atomically:YES encoding:NSUTF8StringEncoding error:nil];
    [controller setValue:path forKey:@"filePath"];
    [controller setValue:@YES forKey:@"dirty"];

    XCTAssertTrue([controller openFile]);

    XCTAssertEqualObjects(self.textView.string, @"# Opened");
    XCTAssertFalse([[controller valueForKey:@"dirty"] boolValue]);
}

- (void)testSaveFileWritesMarkdownAndClearsDirtyState {
    EditorViewController *controller = [self makeControllerWithText:@"# Saved"];
    NSString *path = [self temporaryPathWithName:@"markdown-editor-save.md"];
    [@"old" writeToFile:path atomically:YES encoding:NSUTF8StringEncoding error:nil];
    [controller setValue:path forKey:@"filePath"];
    [controller setValue:@YES forKey:@"dirty"];

    XCTAssertTrue([controller saveFile]);

    NSString *saved = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:nil];
    XCTAssertEqualObjects(saved, @"# Saved");
    XCTAssertFalse([[controller valueForKey:@"dirty"] boolValue]);
}

- (void)testSaveFileRejectsMissingPath {
    EditorViewController *controller = [self makeControllerWithText:@"# Saved"];

    XCTAssertFalse([controller saveFile]);
}

- (void)testMenuActionsUseNonModalBranchesWhenDocumentIsClean {
    TestableEditorViewController *controller = [self makeTestableControllerWithText:@"body"];

    [controller newDocument:nil];
    XCTAssertEqual(controller.newFileCallCount, 1);

    [controller openDocument:nil];
    XCTAssertEqual(self.dialogPresenter.openPanelCallCount, 1);
}

- (void)testNewDocumentDirtySaveSuccessLeavesDocumentInPlace {
    TestableEditorViewController *controller = [self makeTestableControllerWithText:@"body"];
    controller.saveFileResult = YES;
    self.dialogPresenter.confirmationResponse = NSAlertFirstButtonReturn;
    [controller setValue:@YES forKey:@"dirty"];

    [controller newDocument:nil];

    XCTAssertEqual(controller.saveFileCallCount, 1);
    XCTAssertEqual(controller.newFileCallCount, 0);
    XCTAssertEqual(self.dialogPresenter.savePanelCallCount, 0);
}

- (void)testNewDocumentDirtySaveFailureOpensSavePanel {
    TestableEditorViewController *controller = [self makeTestableControllerWithText:@"body"];
    controller.saveFileResult = NO;
    self.dialogPresenter.confirmationResponse = NSAlertFirstButtonReturn;
    [controller setValue:@YES forKey:@"dirty"];

    [controller newDocument:nil];

    XCTAssertEqual(controller.saveFileCallCount, 1);
    XCTAssertEqual(self.dialogPresenter.savePanelCallCount, 1);
}

- (void)testNewDocumentDirtyDiscardCreatesNewFile {
    TestableEditorViewController *controller = [self makeTestableControllerWithText:@"body"];
    self.dialogPresenter.confirmationResponse = NSAlertThirdButtonReturn;
    [controller setValue:@YES forKey:@"dirty"];

    [controller newDocument:nil];

    XCTAssertEqual(controller.newFileCallCount, 1);
}

- (void)testNewDocumentDirtyCancelDoesNothing {
    TestableEditorViewController *controller = [self makeTestableControllerWithText:@"body"];
    self.dialogPresenter.confirmationResponse = NSAlertSecondButtonReturn;
    [controller setValue:@YES forKey:@"dirty"];

    [controller newDocument:nil];

    XCTAssertEqual(controller.newFileCallCount, 0);
    XCTAssertEqual(controller.saveFileCallCount, 0);
}

- (void)testOpenDocumentDirtySaveSuccessLeavesDocumentInPlace {
    TestableEditorViewController *controller = [self makeTestableControllerWithText:@"body"];
    controller.saveFileResult = YES;
    self.dialogPresenter.confirmationResponse = NSAlertFirstButtonReturn;
    [controller setValue:@YES forKey:@"dirty"];

    [controller openDocument:nil];

    XCTAssertEqual(controller.saveFileCallCount, 1);
    XCTAssertEqual(self.dialogPresenter.openPanelCallCount, 0);
}

- (void)testOpenDocumentDirtySaveFailureOpensSavePanel {
    TestableEditorViewController *controller = [self makeTestableControllerWithText:@"body"];
    controller.saveFileResult = NO;
    self.dialogPresenter.confirmationResponse = NSAlertFirstButtonReturn;
    [controller setValue:@YES forKey:@"dirty"];

    [controller openDocument:nil];

    XCTAssertEqual(controller.saveFileCallCount, 1);
    XCTAssertEqual(self.dialogPresenter.savePanelCallCount, 1);
}

- (void)testOpenDocumentDirtyDiscardOpensOpenPanel {
    TestableEditorViewController *controller = [self makeTestableControllerWithText:@"body"];
    self.dialogPresenter.confirmationResponse = NSAlertThirdButtonReturn;
    [controller setValue:@YES forKey:@"dirty"];

    [controller openDocument:nil];

    XCTAssertEqual(self.dialogPresenter.openPanelCallCount, 1);
}

- (void)testOpenDocumentDirtyCancelDoesNothing {
    TestableEditorViewController *controller = [self makeTestableControllerWithText:@"body"];
    self.dialogPresenter.confirmationResponse = NSAlertSecondButtonReturn;
    [controller setValue:@YES forKey:@"dirty"];

    [controller openDocument:nil];

    XCTAssertEqual(self.dialogPresenter.openPanelCallCount, 0);
    XCTAssertEqual(controller.saveFileCallCount, 0);
}

- (void)testSaveMenuActionsUseSaveAndPanelBranches {
    TestableEditorViewController *controller = [self makeTestableControllerWithText:@"body"];
    controller.saveFileResult = YES;

    [controller saveDocument:nil];
    XCTAssertEqual(controller.saveFileCallCount, 1);
    XCTAssertEqual(self.dialogPresenter.savePanelCallCount, 0);

    controller.saveFileResult = NO;
    [controller saveDocument:nil];
    XCTAssertEqual(controller.saveFileCallCount, 2);
    XCTAssertEqual(self.dialogPresenter.savePanelCallCount, 1);

    [controller saveDocumentAs:nil];
    XCTAssertEqual(self.dialogPresenter.savePanelCallCount, 2);
}

- (void)testPromptForOpenAndSaveUseDialogPresenter {
    TestableEditorViewController *controller = [self makeTestableControllerWithText:@"body"];
    NSString *openPath = [self temporaryPathWithName:@"markdown-editor-selected-open.md"];
    NSString *savePath = [self temporaryPathWithName:@"markdown-editor-selected-save.md"];
    self.dialogPresenter.openURL = [NSURL fileURLWithPath:openPath];
    self.dialogPresenter.saveURL = [NSURL fileURLWithPath:savePath];

    [controller promptForOpenURLWithCompletionHandler:nil];
    [controller promptForSaveURLWithCompletionHandler:nil];

    XCTAssertEqual(self.dialogPresenter.openPanelCallCount, 1);
    XCTAssertEqual(self.dialogPresenter.savePanelCallCount, 1);
    XCTAssertEqual(controller.openFileCallCount, 1);
    XCTAssertEqual(controller.saveFileCallCount, 1);
}

- (void)testReloadFileFromDiskIfNeededHonorsDirtyState {
    TestableEditorViewController *controller = [self makeTestableControllerWithText:@"body"];
    [controller setValue:@"/tmp/example.md" forKey:@"filePath"];

    [controller setValue:@NO forKey:@"dirty"];
    [controller reloadFileFromDiskIfNeeded];
    XCTAssertEqual(controller.openFileCallCount, 1);

    [controller setValue:@YES forKey:@"dirty"];
    [controller reloadFileFromDiskIfNeeded];
    XCTAssertEqual(controller.openFileCallCount, 1);
}

- (void)testReplaceCharactersUpdatesTextAndDirtyState {
    EditorViewController *controller = [self makeControllerWithText:@"Hello world"];

    [controller replaceCharactersInRange:NSMakeRange(6, 5) withString:@"Markdown"];

    XCTAssertEqualObjects(self.textView.string, @"Hello Markdown");
    XCTAssertTrue([[controller valueForKey:@"dirty"] boolValue]);
}

- (void)testFormattingActionsUpdateSelectedText {
    EditorViewController *controller = [self makeControllerWithText:@"alpha\nbeta"];

    self.textView.selectedRange = NSMakeRange(0, 5);
    [controller boldButtonClicked:nil];
    XCTAssertEqualObjects(self.textView.string, @"**alpha**\nbeta");

    self.textView.string = @"alpha";
    self.textView.selectedRange = NSMakeRange(0, 5);
    [controller italicButtonClicked:nil];
    XCTAssertEqualObjects(self.textView.string, @"*alpha*");

    self.textView.string = @"alpha";
    self.textView.selectedRange = NSMakeRange(0, 5);
    [controller strikeThroughButtonClicked:nil];
    XCTAssertEqualObjects(self.textView.string, @"~~alpha~~");

    self.textView.string = @"alpha";
    self.textView.selectedRange = NSMakeRange(0, 5);
    [controller codeButtonClicked:nil];
    XCTAssertEqualObjects(self.textView.string, @"```\nalpha\n```");
}

- (void)testBlockFormattingActionsUpdateSelectedText {
    EditorViewController *controller = [self makeControllerWithText:@"alpha\nbeta"];

    self.textView.selectedRange = NSMakeRange(0, self.textView.string.length);
    [controller quoteButtonClicked:nil];
    XCTAssertEqualObjects(self.textView.string, @"> alpha\n> beta\n");

    self.textView.string = @"alpha\nbeta";
    self.textView.selectedRange = NSMakeRange(0, self.textView.string.length);
    [controller listBulletedButtonClicked:nil];
    XCTAssertEqualObjects(self.textView.string, @"- alpha\n- beta\n");

    self.textView.string = @"alpha\nbeta";
    self.textView.selectedRange = NSMakeRange(0, self.textView.string.length);
    [controller listNumberedButtonClicked:nil];
    XCTAssertEqualObjects(self.textView.string, @"1. alpha\n1. beta\n");
}

- (void)testLinkFormattingAndEmptySelection {
    EditorViewController *controller = [self makeControllerWithText:@"alpha"];

    self.textView.selectedRange = NSMakeRange(0, 5);
    [controller insertLinkButtonClicked:nil];
    XCTAssertEqualObjects(self.textView.string, @"[alpha](url)");
    XCTAssertTrue([[controller valueForKey:@"dirty"] boolValue]);

    [controller setValue:@NO forKey:@"dirty"];
    self.textView.selectedRange = NSMakeRange(0, 0);
    [controller applyFormat:TextConverterFormatBold];
    XCTAssertEqualObjects(self.textView.string, @"[alpha](url)");
    XCTAssertFalse([[controller valueForKey:@"dirty"] boolValue]);
}

- (void)testTextDelegateMethodsMarkDocumentDirty {
    EditorViewController *controller = [self makeControllerWithText:@"changed"];

    [controller textDidChange:nil];
    XCTAssertTrue([[controller valueForKey:@"dirty"] boolValue]);

    [controller setValue:@NO forKey:@"dirty"];
    [controller textDidEndEditing:nil];
    XCTAssertTrue([[controller valueForKey:@"dirty"] boolValue]);
}

- (void)testTextViewAllowsAllChanges {
    EditorViewController *controller = [self makeControllerWithText:@"changed"];

    XCTAssertTrue([controller textView:self.textView
              shouldChangeTextInRanges:@[[NSValue valueWithRange:NSMakeRange(0, 1)]]
                     replacementStrings:@[@"x"]]);
}

@end

File: MarkdownEditor/MarkdownEditorTests/EditorViewControllerTests.swift
import Cocoa
import XCTest
@testable import MarkdownEditorLite

final class StubEditorDialogPresenter: EditorDialogPresenter {
    var alertCallCount = 0
    var confirmCallCount = 0
    var openPanelCallCount = 0
    var savePanelCallCount = 0
    var confirmationResponse: NSApplication.ModalResponse = .cancel
    var openURL: URL?
    var saveURL: URL?
    var lastAlertTitle: String?
    var lastAlertMessage: String?

    override func showAlert(title: String, message: String, for window: NSWindow?) {
        alertCallCount += 1
        lastAlertTitle = title
        lastAlertMessage = message
    }

    override func confirmDiscardingChanges(for window: NSWindow?, completionHandler handler: @escaping (NSApplication.ModalResponse) -> Void) {
        confirmCallCount += 1
        handler(confirmationResponse)
    }

    override func showOpenFilePanel(for window: NSWindow?, initialPath: String, completionHandler handler: @escaping (URL?) -> Void) {
        openPanelCallCount += 1
        handler(openURL)
    }

    override func showSaveFilePanel(for window: NSWindow?, initialPath: String, completionHandler handler: @escaping (URL?) -> Void) {
        savePanelCallCount += 1
        handler(saveURL)
    }
}

class TestableEditorViewController: EditorViewController {
    var newFileCallCount = 0
    var saveFileCallCount = 0
    var openFileCallCount = 0
    var saveFileResult = false

    override func newFile() {
        newFileCallCount += 1
    }

    override func saveFile() -> Bool {
        saveFileCallCount += 1
        return saveFileResult
    }

    override func openFile() -> Bool {
        openFileCallCount += 1
        return true
    }
}

final class EditorViewControllerTests: XCTestCase {
    private var textView: NSTextView!
    private var dialogPresenter: StubEditorDialogPresenter!

    private func configureController(_ controller: EditorViewController, text: String) {
        textView = NSTextView(frame: NSRect(x: 0, y: 0, width: 320, height: 200))
        dialogPresenter = StubEditorDialogPresenter()
        textView.string = text
        controller.textView = textView
        controller.dialogPresenter = dialogPresenter
    }

    private func makeController(text: String) -> EditorViewController {
        let controller = EditorViewController()
        configureController(controller, text: text)
        return controller
    }

    private func makeTestableController(text: String) -> TestableEditorViewController {
        let controller = TestableEditorViewController()
        configureController(controller, text: text)
        return controller
    }

    private func temporaryPath(name: String) -> String {
        URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(name).path
    }

    func testDocumentMetadataUsesOutlets() {
        let controller = makeController(text: "body")
        XCTAssertEqual(controller.documentBody(), "body")
    }

    func testDocumentMetadataFallsBackWhenOutletsAreMissing() {
        let controller = EditorViewController()
        XCTAssertEqual(controller.documentBody(), "Unknown")
    }

    func testSetRepresentedObjectIsAccepted() {
        let controller = makeController(text: "body")
        let object = NSObject()

        controller.representedObject = object

        XCTAssertTrue(controller.representedObject as AnyObject === object)
    }

    func testNewFileResetsContentPathAndDirtyState() {
        let controller = makeController(text: "old")
        controller.filePath = "/tmp/example.md"
        controller.dirty = true

        controller.newFile()

        XCTAssertEqual(textView.string, "New File")
        XCTAssertNil(controller.filePath)
        XCTAssertFalse(controller.dirty)
    }

    func testOpenFileRejectsMissingPathsAndDirectories() {
        let controller = makeController(text: "old")

        XCTAssertFalse(controller.openFile())

        controller.filePath = "/tmp/markdown-editor-missing.md"
        XCTAssertFalse(controller.openFile())

        controller.filePath = NSTemporaryDirectory()
        XCTAssertFalse(controller.openFile())
    }

    func testOpenFileLoadsMarkdownAndClearsDirtyState() throws {
        let controller = makeController(text: "old")
        let path = temporaryPath(name: "markdown-editor-open.md")
        try "# Opened".write(toFile: path, atomically: true, encoding: .utf8)
        controller.filePath = path
        controller.dirty = true

        XCTAssertTrue(controller.openFile())

        XCTAssertEqual(textView.string, "# Opened")
        XCTAssertFalse(controller.dirty)
    }

    func testSaveFileWritesMarkdownAndClearsDirtyState() throws {
        let controller = makeController(text: "# Saved")
        let path = temporaryPath(name: "markdown-editor-save.md")
        try "old".write(toFile: path, atomically: true, encoding: .utf8)
        controller.filePath = path
        controller.dirty = true

        XCTAssertTrue(controller.saveFile())

        let saved = try String(contentsOfFile: path, encoding: .utf8)
        XCTAssertEqual(saved, "# Saved")
        XCTAssertFalse(controller.dirty)
    }

    func testSaveFileRejectsMissingPath() {
        let controller = makeController(text: "# Saved")
        XCTAssertFalse(controller.saveFile())
    }

    func testMenuActionsUseNonModalBranchesWhenDocumentIsClean() {
        let controller = makeTestableController(text: "body")

        controller.newDocument(nil)
        XCTAssertEqual(controller.newFileCallCount, 1)

        controller.openDocument(nil)
        XCTAssertEqual(dialogPresenter.openPanelCallCount, 1)
    }

    func testNewDocumentDirtySaveSuccessLeavesDocumentInPlace() {
        let controller = makeTestableController(text: "body")
        controller.saveFileResult = true
        dialogPresenter.confirmationResponse = .alertFirstButtonReturn
        controller.dirty = true

        controller.newDocument(nil)

        XCTAssertEqual(controller.saveFileCallCount, 1)
        XCTAssertEqual(controller.newFileCallCount, 0)
        XCTAssertEqual(dialogPresenter.savePanelCallCount, 0)
    }

    func testNewDocumentDirtySaveFailureOpensSavePanel() {
        let controller = makeTestableController(text: "body")
        controller.saveFileResult = false
        dialogPresenter.confirmationResponse = .alertFirstButtonReturn
        controller.dirty = true

        controller.newDocument(nil)

        XCTAssertEqual(controller.saveFileCallCount, 1)
        XCTAssertEqual(dialogPresenter.savePanelCallCount, 1)
    }

    func testNewDocumentDirtyDiscardCreatesNewFile() {
        let controller = makeTestableController(text: "body")
        dialogPresenter.confirmationResponse = .alertThirdButtonReturn
        controller.dirty = true

        controller.newDocument(nil)

        XCTAssertEqual(controller.newFileCallCount, 1)
    }

    func testNewDocumentDirtyCancelDoesNothing() {
        let controller = makeTestableController(text: "body")
        dialogPresenter.confirmationResponse = .alertSecondButtonReturn
        controller.dirty = true

        controller.newDocument(nil)

        XCTAssertEqual(controller.newFileCallCount, 0)
        XCTAssertEqual(controller.saveFileCallCount, 0)
    }

    func testOpenDocumentDirtySaveSuccessLeavesDocumentInPlace() {
        let controller = makeTestableController(text: "body")
        controller.saveFileResult = true
        dialogPresenter.confirmationResponse = .alertFirstButtonReturn
        controller.dirty = true

        controller.openDocument(nil)

        XCTAssertEqual(controller.saveFileCallCount, 1)
        XCTAssertEqual(dialogPresenter.openPanelCallCount, 0)
    }

    func testOpenDocumentDirtySaveFailureOpensSavePanel() {
        let controller = makeTestableController(text: "body")
        controller.saveFileResult = false
        dialogPresenter.confirmationResponse = .alertFirstButtonReturn
        controller.dirty = true

        controller.openDocument(nil)

        XCTAssertEqual(controller.saveFileCallCount, 1)
        XCTAssertEqual(dialogPresenter.savePanelCallCount, 1)
    }

    func testOpenDocumentDirtyDiscardOpensOpenPanel() {
        let controller = makeTestableController(text: "body")
        dialogPresenter.confirmationResponse = .alertThirdButtonReturn
        controller.dirty = true

        controller.openDocument(nil)

        XCTAssertEqual(dialogPresenter.openPanelCallCount, 1)
    }

    func testOpenDocumentDirtyCancelDoesNothing() {
        let controller = makeTestableController(text: "body")
        dialogPresenter.confirmationResponse = .alertSecondButtonReturn
        controller.dirty = true

        controller.openDocument(nil)

        XCTAssertEqual(dialogPresenter.openPanelCallCount, 0)
        XCTAssertEqual(controller.saveFileCallCount, 0)
    }

    func testSaveMenuActionsUseSaveAndPanelBranches() {
        let controller = makeTestableController(text: "body")
        controller.saveFileResult = true

        controller.saveDocument(nil)
        XCTAssertEqual(controller.saveFileCallCount, 1)
        XCTAssertEqual(dialogPresenter.savePanelCallCount, 0)

        controller.saveFileResult = false
        controller.saveDocument(nil)
        XCTAssertEqual(controller.saveFileCallCount, 2)
        XCTAssertEqual(dialogPresenter.savePanelCallCount, 1)

        controller.saveDocumentAs(nil)
        XCTAssertEqual(dialogPresenter.savePanelCallCount, 2)
    }

    func testPromptForOpenAndSaveUseDialogPresenter() {
        let controller = makeTestableController(text: "body")
        let openPath = temporaryPath(name: "markdown-editor-selected-open.md")
        let savePath = temporaryPath(name: "markdown-editor-selected-save.md")
        dialogPresenter.openURL = URL(fileURLWithPath: openPath)
        dialogPresenter.saveURL = URL(fileURLWithPath: savePath)

        controller.promptForOpenURL(completionHandler: nil)
        controller.promptForSaveURL(completionHandler: nil)

        XCTAssertEqual(dialogPresenter.openPanelCallCount, 1)
        XCTAssertEqual(dialogPresenter.savePanelCallCount, 1)
        XCTAssertEqual(controller.openFileCallCount, 1)
        XCTAssertEqual(controller.saveFileCallCount, 1)
    }

    func testReloadFileFromDiskIfNeededHonorsDirtyState() {
        let controller = makeTestableController(text: "body")
        controller.filePath = "/tmp/example.md"

        controller.dirty = false
        controller.reloadFileFromDiskIfNeeded()
        XCTAssertEqual(controller.openFileCallCount, 1)

        controller.dirty = true
        controller.reloadFileFromDiskIfNeeded()
        XCTAssertEqual(controller.openFileCallCount, 1)
    }

    func testReplaceCharactersUpdatesTextAndDirtyState() {
        let controller = makeController(text: "Hello world")

        controller.replaceCharacters(in: NSRange(location: 6, length: 5), with: "Markdown")

        XCTAssertEqual(textView.string, "Hello Markdown")
        XCTAssertTrue(controller.dirty)
    }

    func testFormattingActionsUpdateSelectedText() {
        let controller = makeController(text: "alpha\nbeta")

        textView.setSelectedRange(NSRange(location: 0, length: 5))
        controller.boldButtonClicked(nil)
        XCTAssertEqual(textView.string, "**alpha**\nbeta")

        textView.string = "alpha"
        textView.setSelectedRange(NSRange(location: 0, length: 5))
        controller.italicButtonClicked(nil)
        XCTAssertEqual(textView.string, "*alpha*")

        textView.string = "alpha"
        textView.setSelectedRange(NSRange(location: 0, length: 5))
        controller.strikeThroughButtonClicked(nil)
        XCTAssertEqual(textView.string, "~~alpha~~")

        textView.string = "alpha"
        textView.setSelectedRange(NSRange(location: 0, length: 5))
        controller.codeButtonClicked(nil)
        XCTAssertEqual(textView.string, "```\nalpha\n```")
    }

    func testBlockFormattingActionsUpdateSelectedText() {
        let controller = makeController(text: "alpha\nbeta")

        textView.setSelectedRange(NSRange(location: 0, length: (textView.string as NSString).length))
        controller.quoteButtonClicked(nil)
        XCTAssertEqual(textView.string, "> alpha\n> beta\n")

        textView.string = "alpha\nbeta"
        textView.setSelectedRange(NSRange(location: 0, length: (textView.string as NSString).length))
        controller.listBulletedButtonClicked(nil)
        XCTAssertEqual(textView.string, "- alpha\n- beta\n")

        textView.string = "alpha\nbeta"
        textView.setSelectedRange(NSRange(location: 0, length: (textView.string as NSString).length))
        controller.listNumberedButtonClicked(nil)
        XCTAssertEqual(textView.string, "1. alpha\n1. beta\n")
    }

    func testLinkFormattingAndEmptySelection() {
        let controller = makeController(text: "alpha")

        textView.setSelectedRange(NSRange(location: 0, length: 5))
        controller.insertLinkButtonClicked(nil)
        XCTAssertEqual(textView.string, "[alpha](url)")
        XCTAssertTrue(controller.dirty)

        controller.dirty = false
        textView.setSelectedRange(NSRange(location: 0, length: 0))
        controller.applyFormat(.bold)
        XCTAssertEqual(textView.string, "[alpha](url)")
        XCTAssertFalse(controller.dirty)
    }

    func testTextDelegateMethodsMarkDocumentDirty() {
        let controller = makeController(text: "changed")

        controller.textDidChange(Notification(name: NSText.didChangeNotification))
        XCTAssertTrue(controller.dirty)

        controller.dirty = false
        controller.textDidEndEditing(Notification(name: NSText.didEndEditingNotification))
        XCTAssertTrue(controller.dirty)
    }

    func testTextViewAllowsAllChanges() {
        let controller = makeController(text: "changed")

        XCTAssertTrue(controller.textView(
            textView,
            shouldChangeTextInRanges: [NSValue(range: NSRange(location: 0, length: 1))],
            replacementStrings: ["x"]
        ))
    }
}
